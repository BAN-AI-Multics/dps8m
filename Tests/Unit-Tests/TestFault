do ../../src/dps8/base_system.ini
load ./TestFault.oct 
set debug stdout
set cpu debug=trace;appending;regdump;addrmod;fault
;set cpu debug=trace
;set cpu debug=err
g 0
q

