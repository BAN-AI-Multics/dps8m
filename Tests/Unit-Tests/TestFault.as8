"
" Access an unmapped segment page; fault; map the page; return from fault, 
" restarting the faulted instruction.
"
" Virtual Memory layout
"
"     segment
"       (0)      absolute mode code
"        1       executable code and data
"        2       its/itp data
"
	equ	seg1,1
	equ	seg2,2

"     segment 1 pages
"
"       0-n    code
"       1000   data
"       1001   data
"       1002   data
"       1003   data

	equ	pgsz,1024	" pages are 1024 words long
	equ	program_start,100*pgsz

"""
""" Entry point
"""

	org	0
	tra	start


	org	program_start

ldbr_data:

"""
""" DSBR format for the LDBR instruction:
"""

"        pos len
"  ADDR    0  24  If U = 1, the memory address of the origin of the descriptor
"                 segment; if U = 0, the memory address of the descriptor
"                 segment page table
" 
"  BOUND  37  14  The 14 most significant bits of the highest Y-block16 
"                 address of the descriptor segment that can be addressed 
"                 without causing an fault.

"  U      55   1  U = 0: descriptor segment is paged; 1: unpaged.
"  STACK  60  12  Used by CALL6; the upper 12 bits of the 15 bit stack base
"                 segment number.
"
" U bit
	equ	paged,0
	equ	unpaged,1

" F bit
	equ	nonresident,0
	equ	resident,1

        equ	dsbr_bound,0	" All of our segments will fit in a single
				" Y16 (2 words/segment; this allows 8)

	vfd	24/descr_seg_page_table,12/0
	vfd	1/0,14/dsbr_bound,4/0,1/paged,4/0,12/0

"""
""" Descriptor Segment Page Table.
"""


descr_seg_page_table:

" A table of PTWs; each PTW points to a table of 1024 SDWs; the first PTW is
" used for segment numbers 0-1023, the second for 1024-2047, and so on
"
" In this table, each PTW takes up 2 words; the second word is unused.
" (If the DSPT is unpaged, it is an array of SDWs, which are two words.)

" PTW format:
"    ADDR  x x x x x x 0 0 U 0 0 M x x x F FC
"    x x x x x x x x x x x x x x x x x x x x x x x x x x x x x x x x x x x x x 
"
"        pos len
"  ADDR:   0  18  modulo 64 absolute main memory address of page
"     U   26   1  page has been used
"     M   29   1  page has been modfied
"     F   33   1  0: page not in memory; 1: page in memory
"     FC  34   2  direct fault number
   
" The following vfd allocate 24 bits for the addr to hand-wave away the 
" divide by 64.

" segments 0-1023

	vfd	24/seg_0_descr_seg_page_table,2/0,1/0,2/0,1/0,3/0,1/resident,2/0
	vfd	36/0

"""
""" SDW format
"""

" even word
"        pos len
" ADDR     0  24   U=1: address of the unpaged segment; 0 address of the 
"                  segment page table
" R1      24   3
" R2      27   3
" R3      30   3
" F       33   1    0: page is non-resident, 1: resident
" FC      34   2

" odd word
"        pos len
" BOUND    1  14
" R       15   1
" E       16   1
" W       17   1
" P       18   1
" U       19   1   0: segment is paged; 1: unpaged
" G       20   1
" C       21   1
" ER      22  14

	sixtyfour

"""
""" Descriptor segment page for segments 0-1023
""""

seg_0_descr_seg_page_table:

" segment 0 (non-existant)
	vfd	24/-1, 3/0, 3/0, 3/0, 1/nonresident, 2/0
	vfd	1/0,14/0,1/0,1/0,1/0,1/0,1/paged,1/0,1/1,14/0

" segment 1; code and data R/E/W
"	vfd	24/seg_1_page_table,3/5,3/5,3/5,1/resident,2/3
	vfd	24/seg_1_page_table,3/0,3/0,3/0,1/resident,2/3
	vfd	1/0,14/-1,1/1,1/1,1/1,1/0,1/paged,1/0,1/1,14/0

" segment 2; data R/W
"	vfd	24/seg_2_page_table,3/5,3/5,3/5,1/resident,2/3
	vfd	24/seg_2_page_table,3/0,3/0,3/0,1/resident,2/3
	vfd	1/0,14/-1,1/1,1/0,1/1,1/0,1/paged,1/0,1/1,14/0


" 1024 segment, 3 segs used, 2 words/seg
"   (1024 - 3) * 2
        bss	2042

seg_1_page_table:

	vfd	24/seg_1_page_0,2/0,1/0,2/0,1/0,3/0,1/resident,2/0

seg_2_page_table:

	vfd	24/seg_2_page_0,2/0,1/unpaged,2/0,1/0,3/0,1/resident,2/0


"""
""" Abs mode code
"""

	sixtyfour
start:	
	eaa	msg_main_entry
	emcall	16

	ldbr	ldbr_data

	lprp0	entry_ptr
	tra	0|0

	oct	its,43
entry_ptr:
	vfd	6/0,12/1,18/main-seg_1_page_0

msg_main_entry:
	aci	'Fault test\n\0\0'


"""
""" Segment 1
"""

	sixtyfour
seg_1_page_0:

main:
	eaa	msg_main_entry
	emcall	16
	emcall	18

"""
""" Segment 2
"""

	sixtyfour
seg_2_page_0:

