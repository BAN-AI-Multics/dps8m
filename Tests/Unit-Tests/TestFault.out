
dps-8/m simulator V4.0-0 Beta        git commit id: c8ae76cb
310 (000466) words loaded
477 (000735) words loaded
3 (000003) words loaded
Debug output to "STDOUT"
Debug output to "STDOUT" at Wed Aug  6 09:40:22 2014
Debug set to break at cycle: 10000
Fault test entry
Fault test main
Running tests without page faulting
Test 1 R,n ...
... ok
Test 2 R,* ...
... ok
Test 3 R,*,* ...
... ok
Test 4 R,3* ...
... ok
Test 5 R,*seg ...
... ok
Test 6 R,*seg, * ...
... ok
Test 7 R,sd ...
... ok
Test 8 R,scr ...
... ok
Test 9 R,scr w/addr change ...
... ok
Test 10 R,ci ...
... ok
Test 11 R,i ...
... ok
Test 12 R,sc ...
... ok
Test 13 R,sc w/addr change ...
... ok
Test 14 R,ad ...
... ok
Test 15 R,di ...
... ok
Test 16 R,dic ...
... ok
Test 17 R,id ...
... ok
Test 18 R,idc ...
... ok
Test 19 R,*3 ...
... ok
Running tests with page faulting
Test 1 R,n ...
[1:000000 000024](1:000000 006000)... ok
Test 2 R,* ...
[1:000000 000037](1:000000 002000)(1:000000 006001)... ok
Test 3 R,*,* ...
[1:000000 000052](1:000000 002001)(1:000000 004000)(1:000000 006002)... ok
Test 4 R,3* ...
[1:000000 000066](1:000000 002003)(1:000000 006003)... ok
Test 5 R,*seg ...
[1:000000 000101](1:000000 002004)(2:000000 002000)(1:000000 006004)... ok
Test 6 R,*seg, * ...
[1:000000 000114](1:000000 002006)(2:000000 000000)(2:000000 002001)(1:000000 006005)... ok
Test 7 R,sd ...
[1:000000 000131](1:000000 002010)(1:000000 006006)... ok
Test 8 R,scr ...
[1:000000 000146](1:000000 002011)(1:000000 006010)... ok
Test 9 R,scr w/addr change ...
[1:000000 000163](1:000000 002012)(1:000000 006011)... ok
Test 10 R,ci ...
[1:000000 000200](1:000000 002013)(1:000000 006013)... ok
Test 11 R,i ...
[1:000000 000213](1:000000 002014)(1:000000 006014)... ok
Test 12 R,sc ...
[1:000000 000230](1:000000 002015)(1:000000 006015)... ok
Test 13 R,sc w/addr change ...
[1:000000 000245](1:000000 002016)(1:000000 006016)[1:000000 000253](1:000000 002016)(1:000000 006017)... ok
Test 14 R,ad ...
[1:000000 000270](1:000000 002017)(1:000000 006020)[1:000000 000276](1:000000 002017)(1:000000 006021)... ok
Test 15 R,di ...
[1:000000 000313](1:000000 002020)(1:000000 006022)... ok
Test 16 R,dic ...
[1:000000 000330](1:000000 002021)(1:000000 004001)(1:000000 006024)... ok
Test 17 R,id ...
[1:000000 000345](1:000000 002022)(1:000000 006025)... ok
Test 18 R,idc ...
[1:000000 000362](1:000000 002023)(1:000000 004003)(1:000000 006026)... ok
Test 19 R,*3 ...
[1:000000 000376](1:000000 002024)(1:000000 006030)... ok
Fault test done

simCycles = 7810

cpuCycles = 3863
Master mode entry 1 faults = 42
Directed fault 3 faults = 68
TestFault-17> g 0

Halt, IC: 000016 (000000000000)
Goodbye
Debug output disabled
